doc/mgmt/gnmi/master_arbitration.md

```diff
@@ -0,0 +1,331 @@
+# gNMI Master Arbitration HLD
+
+## Table of Content
+- [Revision](#revision)
+- [Scope](#scope)
+- [Definitions / Abbreviations](#definitions--abbreviations)
+- [Overview](#overview)
+- [Requirements](#requirements)
+- [Architecture Design](#architecture)
+- [High-Level Design](#high-level-design)
+- [SAI API](#sai-api)
+- [Configuration and management](#configuration-and-management)
+- [Warmboot and Fastboot Design Impact](#warmboot-and-fastboot-design-impact)
+- [Restrictions / Limitations](#restrictions--limitations)
+- [Testing Requirements / Design](#testing-requirements--design)
+- [Open / Action items - if any](#open--action-items---if-any)
+
+### Revision
+
+| Rev  | Rev Date   | Author(s)          | Change Description |
+|------|------------|--------------------|--------------------|
+| v0.1 | 03/14/2023 | Tomek Madejski (Google), Sally Lyu (Google) | Initial version |
+
+### Scope
+
+This document describes the high level design of gNMI Master Arbitration. It provides a detailed explanation of the definition, as well as the implementation needed in the SONIC framework.
+
+
+### Definitions/Abbreviations
+
+gNMI - gRPC Network Management Interface (gNMI) is a protocol to manage and report a network element's configuration and operational data.
+
+gNMI client - a software application that is used to connect to a network element and manage its configuration and operational data. The gNMI client can be written in any language that supports gRPC.
+
+SDN - Software-defined networking (SDN) is a network architecture that uses software-based controllers or application programming interfaces (APIs) to communicate with underlying hardware infrastructure and direct traffic on a network.
+
+EID -  Election ID. A 128-bit integer. The election ID MUST be guaranteed to be monotonically increasing.But how it is done is out of scope of this document and implementation.
+
+Reference: https://github.com/openconfig/reference/blob/master/rpc/gnmi/gnmi-master-arbitration.md
+
+### Overview
+
+For high availability, a system may run multiple replicas of a gNMI client (controller). Among the replicas, only one client should be elected as master and do gNMI operations that mutate any state on the target. However, in the event of a network partition, there can be two or more replicas thinking themselves as master. For read-only RPCs, this is OK. But if they both call the `Set` RPC, the target may be incorrectly configured by the stale master. Therefore, "Master Arbitration" is needed when multiple clients exist.
+
+**Definition**
+
+When Master Arbitration feature is turned on, the gNMI server receives a proto with the election ID (EID) specified in it. The election ID is a 128-bit integer MUST be guaranteed to be monotonically increasing -- a new master's election ID MUST always be larger than its predecessor. The process through which election IDs are allocated to client replicas is out-of-scope of this specification.
```

Comment:
"a new master's election ID MUST always be larger than its predecessor"
In theory, on the roll over of the 128-bit integer, wouldn't this assumption be breached ?

Reply:
The implementation of this feature is as designed and described in the spec [link](https://github.com/openconfig/reference/blob/master/rpc/gnmi/gnmi-master-arbitration.md). Assignment of ID is out of switch control. The controller designers have to handle this problem by, say, restarting the switch before the ID rolls over, and make sure the ID is unique.

Reply:
Not sure restarting the switch before the ID rolls over is the good idea, do we have any mechanism to validate whether the switches are using the right EID, also the switches are using the right master controller.

Reply:
I don't think there is a `RIGHT` EID in the case. I understand your concern. But this can only be addressed outside of switch (EID manager should do the job). Switch side has no knowledge to make sure the `RIGHT` controller is sending request.